common/djangoapps/edxmako/template.py

```diff
@@ -13,47 +13,89 @@
+    be rendered like it is a Django template because the arguments are transformed
     in a way that MakoTemplate can understand.
     """
 
     def __init__(self, *args, **kwargs):
         """Overrides base __init__ to provide django variable overrides"""
-        if not kwargs.get('no_django', False):
-            kwargs['lookup'] = edxmako.LOOKUP['main']
-        super(Template, self).__init__(*args, **kwargs)
+        self.engine = kwargs.pop('engine', engines[Engines.MAKO])
+        if len(args) and isinstance(args[0], MakoTemplate):
+            self.mako_template = args[0]
+        else:
+            kwargs['lookup'] = LOOKUP['main']
+            self.mako_template = MakoTemplate(*args, **kwargs)
 
-    def render(self, context_instance):
+    def render(self, context=None, request=None):
         """
         This takes a render call with a context (from Django) and translates
         it to a render call on the mako template.
         """
-        # collapse context_instance to a single dictionary for mako
-        context_dictionary = {}
+        context_object = self._get_context_object(request)
+        context_dictionary = self._get_context_processors_output_dict(context_object)
+
+        if isinstance(context, Context):
+            context_dictionary.update(context.flatten())
+        elif context is not None:
+            context_dictionary.update(context)
+
+        self._add_core_context(context_dictionary)
+        self._evaluate_lazy_csrf_tokens(context_dictionary)
+
+        return self.mako_template.render_unicode(**context_dictionary)
+
+    @staticmethod
+    def _get_context_object(request):
+        """
+        Get a Django RequestContext or Context, as appropriate for the situation.
+        In some tests, there might not be a current request.
+        """
+        request_context = get_template_request_context(request)
+        if request_context is not None:
+            return request_context
+        else:
+            return Context({})
 
-        # In various testing contexts, there might not be a current request context.
-        request_context = get_template_request_context()
-        if request_context:
-            for item in request_context:
-                context_dictionary.update(item)
-        for item in context_instance:
-            context_dictionary.update(item)
+    def _get_context_processors_output_dict(self, context_object):
+        """
+        Run the context processors for the given context and get the output as a new dictionary.
+        """
+        with context_object.bind_template(self):
+            return context_object.flatten()
+
+    @staticmethod
+    def _add_core_context(context_dictionary):
+        """
+        Add to the given dictionary context variables which should always be
+        present, even when context processors aren't run during tests.  Using
+        a context processor should almost always be preferred to adding more
```

Comment:
Nit: Perhaps stronger language here? Such as: "Never add to these variables."?

Reply:
Not sure if that's appropriate; I think the most accurate guidance is "only add things here needed to render the base templates", so this already feels a little on the scare words side to me.